### ISSUE TITLE
include_directories doesn't work for Vala

### ISSUE THREAD

**Issue by zeenix** [mesonbuild/meson#906]
Currently you have to manually specify VAPI paths with vala_args. include_directories should take care of that.

**Comment by nirbheek** [comment]
I don't think it makes sense for `include_directories` to work with `.vapi` files. `.vapi` files are kind-of like headers, but the way `valac` consumes Vapis is more like how `ld` consumes libraries: you can either specify the full path to it on the command line, or you can specify the name and search directories.

The way people should use Vapis in build targets is the same as C libraries:

1) System-installed vapis: `dependency()` (if there's a pkg-config file for it) or `valac.find_library()` (to be implemented) and use with the `dependencies :` kwarg to the build target.

2) Generated from a GIR using [`gnome.generate_vapi`](https://github.com/mesonbuild/meson/pull/895): Yields an `InternalDependency` object that can be used in the same way as (1) (identical to `declare_dependency` in C-like).

3) Generated by Valac when creating a `library()`: Use the return value as a library object, for example:

```
somelib = library('somelib.vala', 'config.vala')
executable('vala-app', 'app.vala', link_with : d)
```
If you want to link a C executable to this library, you will have to use `declare_dependency` which should add the include path to the C header generated by `valac`:
```
somelib = library('somelib.vala', 'config.vala')
d = declare_dependency(link_with: somelib)
executable('vala-app', 'app.vala', dependencies : d)
executable('c-app', 'app.c', dependencies : d)
```
Note that currently the vapi name is the same as the library name. #912 is a PR to add the ability to specify it with kwargs to `library()`.

4) Hand-written (with the sources) pointing to system-installed C libraries: Either add it to the list of sources for the build target like a source file, or if you want to expose it to subprojects, use `declare_dependency()` without a `link_with :` kwarg, and instead just add it to the `sources :` kwarg. This will also work with the `dependencies :` kwarg to a build target.

If this is not how things work, we should fix those bugs. Can you please try this with the vapis used in Boxes and get back to us?

**Comment by nirbheek** [comment]
(1) and (2) will be shipped in 0.36.0. (3) requires more changes that will go into 0.37.0

**Comment by nirbheek** [comment]
(3) has been implemented, and so has `valac.find_library()`. This issue is resolved.

Some further issues of interest are #1194 and #1195
